lib/buffer.js

```diff
@@ -256,7 +256,7 @@ function assertSize(size) {
 
   if (typeof size !== 'number') {
     err = new ERR_INVALID_ARG_TYPE('size', 'number', size);
-  } else if (size < 0 || size > kMaxLength) {
+  } else if (size < 0 || size > kMaxLength || Number.isNaN(size)) {
```

Comment:
Bringing in `Number.isNan()` here adds unnecessary overhead:

1. An additional branch instruction
2. An additional function call

This can be avoided by changing the condition to this instead:

`} else if (!(size >= 0 && size <= kMaxLength)) {`

This is sufficient to catch `NaN`, as well as anything else that's not a true number.

Reply:
Otherwise, the range check is still not right.

Reply:
It seems that both if fine.

> Otherwise, the range check is still not right.

I think this is identical to `} else if (!(size >= 0 && size <= kMaxLength)) {` in behavior, could you please tell me why the range check is still not right ?

Reply:
> It seems that both [are] fine.

The one is unnecessarily expensive (extra branch, extra function call).

> I think this is identical

No, the original checks whether any type of variable is < 0 or > max, and only throws if one of those is true.

This is not logically the same as checking whether any type of variable is exactly >= 0 && <= max, and throwing if not.

For example:

The former doesn't test for membership inside the range, only for potential membership outside the range (depending on type). So variables like `NaN` pass and are considered within the range, even when they are not.

The latter is a much stricter test, which variables such as `NaN` do not pass.

Reply:
> The latter is a much stricter test, which variables such as `NaN` do not pass.

You mentioned that a few times, but what *are* other number values besides `NaN` for which this makes a difference? I would have been pretty sure there are any?

Reply:
(i.e. I agree that the extra function call might be unnecessary, but the code *does* seem correct.)

Reply:
@addaleax , the point is that if we had had the stricter "within range" test to begin with, we wouldn't have had this issue with NaN. We shouldn't keep going down this road of asking what types could possibly fall outside the range. That's the blacklist approach. Node should be using the whitelist approach for these kinds of checks.

But to answer your question, I think NaN might just be the last of the gang to die.